buildSrc/src/main/java/xyz/wagyourtail/doclet/tsdoclet/Main.java

```diff
@@ -90,34 +96,81 @@ public boolean run(DocletEnvironment environment) {
+                """
+                declare const event: Events[keyof Events];
                 declare const file: _javatypes.java.io.File;
-                declare const context: _javatypes.xyz.wagyourtail.jsmacros.core.language.EventContainer<any>;
+                declare const context: EventContainer;
 
                 declare namespace Events {
-                    export interface BaseEvent extends _javatypes.java.lang.Object {
-                        getEventName(): string;
-                    }""");
 
+                    export interface BaseEvent extends JavaObject {
+                        getEventName(): string;
+                    }"""
+            );
             for (EventParser event : eventClasses) {
-                outputTS.append(StringHelpers.tabIn("\n\n" + event.genTSInterface()));
+                outputTS.append("\n\n" + StringHelpers.tabIn(event.genTSInterface()));
             }
 
-            outputTS.append("\n}");
+            outputTS.append("\n\n}\n\ninterface Events {\n");
+            for (EventParser event : eventClasses) {
+                outputTS.append("\n    ").append(event.getName())
+                    .append(": Events.").append(event.getName()).append(";");
+            }
+            outputTS.append("\n\n}");
 
             for (LibraryParser lib : libraryClasses) {
                 outputTS.append("\n\n" + lib.genTSInterface());
             }
 
             outputTS.append("\n\ndeclare " + classes.genTSTree());
-            outputTS.append("\n\ndeclare namespace Java {");
+            outputTS.append("\n\ninterface JavaTypeDict {");
             for (ClassParser clz : classes.getAllClasses()) {
-                outputTS.append("\n").append("    export function type(className: \"").append(clz.getType().getQualifiedName().toString()).append("\"): ")
-                        .append("_javatypes.java.lang.Class<_javatypes.").append(clz.getType().getQualifiedName().toString().replace(".function.", "._function."))
-                        .append("> & _javatypes.").append(clz.getType().getQualifiedName().toString().replace(".function.", "._function.")).append(".static");
+                outputTS.append("\n    \"").append(clz.getTypeString().replaceFirst("<.+$", "").replace("_javatypes.", "")).append("\": ")
+                    .append("JavaClass<").append(clz.getShortifiedType().replaceFirst("<.+$", "").replaceFirst("^\\$", "").replace(".function.", "._function."))
+                    .append("> & ").append(clz.getTypeString().replaceFirst("<.+$", "").replace(".function.", "._function.")).append(".static;");
+            }
+            outputTS.append("\n}\n\ntype _ = { [none: symbol]: never }; // to trick vscode to rename types\n");
+
+            int maxLen = 0;
+            int maxRedirLen = 0;
+            for (ClassParser clz : classes.getAllClasses()) { // count
+                String type = clz.getTypeString();
+                if (!type.startsWith("_javatypes.xyz.")) continue;
```

Comment:
maybe we limit it to `xyz.wagyourtail.jsmacros` tho I suppose the docs don't run with any other xyz namespaces anyway...

Reply:
i'll try, maybe tomorrow

Reply:
actually there's 8 namespaces in _javatypes
com, io, it, java, javassist, javax, org, xyz
continue?

Reply:
done b9ae025